plugin/src/main/scala/chisel3/internal/plugin/IdentifierComponent.scala

```diff
@@ -0,0 +1,122 @@
+    def isExactBaseModule(sym: Symbol): Boolean = { sym.tpe =:= baseModuleTpe }
+
+    def isNullaryMethodNamed(name: String, defdef: DefDef): Boolean =
+      defdef.name.decodedName.toString == name && defdef.tparams.isEmpty && defdef.vparamss.isEmpty
+
+    def getConstructorAndParams(body: List[Tree]): (Option[DefDef], Seq[Symbol]) = {
+      val paramAccessors = mutable.ListBuffer[Symbol]()
+      var primaryConstructor: Option[DefDef] = None
+      body.foreach {
+        case acc: ValDef if acc.symbol.isParamAccessor && !acc.mods.hasFlag(Flag.BYNAMEPARAM) =>
+          paramAccessors += acc.symbol
+        case con: DefDef if con.symbol.isPrimaryConstructor =>
+          primaryConstructor = Some(con)
+        case d: DefDef if isNullaryMethodNamed("_moduleDefinitionIdentifierProposal", d) =>
+          val msg = "Users cannot override _moduleDefinitionIdentifierProposal. Let the compiler plugin generate it."
+          global.reporter.error(d.pos, msg)
+        case _ =>
+      }
+      (primaryConstructor, paramAccessors.toList)
+    }
+
+    def generateIdentifierMethod(module: ClassDef, thiz: global.This, baseClass: global.Symbol): Tree = {
+      val (conOpt, params) = getConstructorAndParams(module.impl.body)
+
+      // The params have spaces after them (Scalac implementation detail)
+      val paramLookup: Map[String, Symbol] = params.map { sym => sym.name.toString.trim -> sym }.toMap
+
+      val str = stringFromTypeName(module.name)
+      // Create a getProposal(this.<ref>) for each field matching order of constructor arguments
+      val tpedNames: List[Tree] = (localTyper.typed(q"$str")) +:
+        conOpt.toList.flatMap { x =>
+          x.vparamss.flatMap(_.flatMap { vp =>
+            paramLookup.get(vp.name.toString) match {
+              case Some(p) =>
+                // Make this.<ref>
+                val select = gen.mkAttributedSelect(thiz.asInstanceOf[Tree], p)
+                List(localTyper.typed(q"_root_.chisel3.naming.IdentifierProposer.getProposal($select)"))
+              case None => Nil
+            }
+          })
+        }
+      val body = localTyper.typed(q"_root_.chisel3.naming.IdentifierProposer.makeProposal(..$tpedNames)")
+
+      // Create the symbol for the method and have it be associated with the Module class
+      val identifierSym = module.symbol.newMethod(
+        TermName("_moduleDefinitionIdentifierProposal"),
+        module.symbol.pos.focus,
+        Flag.OVERRIDE | Flag.PROTECTED
+      )
+      identifierSym.setInfo(NullaryMethodType(stringTpe))
+
+      localTyper.typed(DefDef(identifierSym, body)).asInstanceOf[DefDef]
+    }
+
+    override def transform(tree: Tree): Tree = tree match {
+
+      case module: ClassDef
+          if isAModule(module.symbol)
+            && !isExactBaseModule(module.symbol)
+            && !module.mods.hasFlag(Flags.TRAIT)
+            && !module.name.decode.contains("$anon") =>
```

Comment:
Does this not work on anonymous functions?

Reply:
Anonymous classes. There doesn't exist a new name for `new MyModule {}` that is better that whatever was derived for `MyModule`, so there is no point in overriding the `_moduleDefinitionIdentifierProposal` for this anonymous class.

Reply:
Is there a better way to detect it than String matching the name?

Reply:
Not that I can figure out, sadly.